slick-hikaricp/src/main/scala/slick/jdbc/hikaricp/HikariCPJdbcDataSource.scala

```diff
@@ -1,16 +1,69 @@
 package slick.jdbc.hikaricp
 
-import java.sql.{Driver, Connection}
+import java.sql.{Connection, Driver}
+import java.util.concurrent.atomic.AtomicInteger
+import java.util.concurrent.locks.ReentrantLock
+
 import com.typesafe.config.Config
 import slick.SlickException
-import slick.jdbc.{JdbcDataSourceFactory, JdbcDataSource}
+import slick.jdbc.{JdbcDataSource, JdbcDataSourceFactory}
 import slick.util.ConfigExtensionMethods._
+import slick.util.Logging
 
 /** A JdbcDataSource for a HikariCP connection pool.
   * See `slick.jdbc.JdbcBackend#Database.forConfig` for documentation on the config parameters. */
-class HikariCPJdbcDataSource(val ds: com.zaxxer.hikari.HikariDataSource, val hconf: com.zaxxer.hikari.HikariConfig) extends JdbcDataSource {
-  def createConnection(): Connection = ds.getConnection()
+class HikariCPJdbcDataSource(val ds: com.zaxxer.hikari.HikariDataSource, val hconf: com.zaxxer.hikari.HikariConfig)
+  extends JdbcDataSource
+  with Logging {
+
+  /**
+    * the poolLock must be obtained before acquiring/releasing Connections
+    */
+  private[this] val poolLock: ReentrantLock = new ReentrantLock()
+  private[this] val poolNotFullCondition = poolLock.newCondition()
+  private[this] var connectionsInUseCount: Int = 0 //should only be accessed after obtaining the poolLock
+
+  def createConnection(): Connection = {
+    poolLock.lock()
+    try {
+      if (logger.isDebugEnabled) logger.debug("number of connections in use = "+connectionsInUseCount)
+      while (connectionsInUseCount == hconf.getMaximumPoolSize) {
+          poolNotFullCondition.await()
+      }
+      val c = new DelegateConnection(ds.getConnection()) {
+        override def close(): Unit = {
+          closeConnection(conn)
+        }
+      }
+      connectionsInUseCount += 1
+      if (connectionsInUseCount == hconf.getMaximumPoolSize) {
+        logger.debug("Maximum number of connections in use reached, pausing")
+        pause()
+      }
+      c
+    } finally {
+      poolLock.unlock()
+    }
+  }
+
+  private[this] def closeConnection(connection: Connection) = {
+    poolLock.lock()
+    try {
+      connection.close()
+    } finally {
+      connectionsInUseCount -= 1
+      if (connectionsInUseCount == hconf.getMaximumPoolSize -1) {
+        logger.debug("Connection freed below maximum, resuming")
+        poolNotFullCondition.signal()
+        resume()
+      }
+      poolLock.unlock()
```

Comment:
unlocking always needs to go in their own finallies, to avoid production deadlocks.

~~~scala
 +        try {
 +          connectionsInUseCount -= 1
 +        if (connectionsInUseCount == hconf.getMaximumPoolSize -1) {
 +          logger.debug("Connection freed below maximum, resuming")
 +          poolNotFullCondition.signal()
 +          resume()
 +        }
+       } finally poolLock.unlock()
~~~